Break the dependency of `:detekt-api` with `:detekt-psi-utils`

## Expected Behavior
`:detekt-api` shouldn't depend on `:detekt-psi-utils`.

## Current Behavior
It does

## Context
This idea comes from this comment: https://github.com/detekt/detekt/pull/7105#issuecomment-2034125369

I did a fast check and the main reason to have this dependency is because `:detekt-api` uses `FilePath` and that class is declared on `:detekt-psi-utils`. That have little sense, that class should be defined on `:detekt-api`. But once you move that class to `:detekt-api` a lot of `:detekt-psi-utils` doesn't compile because they use `FilePath`. So maybe the solution is to invert the dependency, I'm not sure.

Also this could help to unblock #7123.

We should actually go further, but I don't think it's possible to solve while we support autocorrect.

If we created each `KtFile` the proper way the absolute file path would be available at `KtFile.virtualFile.path` and we wouldn't need to store it again. The relative path can be generated from this & the base path. This is actually how we get the absolute path with the compiler plugin:
https://github.com/detekt/detekt/blob/68b949f3ed9965a03797d2951cce68fd7dfc87a2/detekt-psi-utils/src/main/kotlin/io/github/detekt/psi/KtFiles.kt#L33-L37

Why is it available in compiler plugin, but not when we generate them? Because we use a different method to create each `KtFile`. detekt uses `createPhysicalFile` here:
https://github.com/detekt/detekt/blob/68b949f3ed9965a03797d2951cce68fd7dfc87a2/detekt-parser/src/main/kotlin/io/github/detekt/parser/KtCompiler.kt#L34-L37

While the Kotlin compiler will add the source files to the KotlinCoreEnvironment then use `KotlinCoreEnvironment.getSourceFiles()` which returns `List<KtFile>`.

Why can't we do the same? Because `KtFile` created by the compiler are not writeable, but files created by the KtPsiFileFactory are. And files have to be writeable for autocorrect to work, at least how it's currently implemented, otherwise we would get errors like:

```
Caused by: org.jetbrains.kotlin.com.intellij.util.IncorrectOperationException: Cannot modify a read-only file 'D:\IdeaProjects\detekt\detekt-api\src\main\kotlin\io\gitlab\arturbosch\detekt\api\internal\PathFilters.kt'.
	at org.jetbrains.kotlin.com.intellij.psi.impl.CheckUtil.checkWritable(CheckUtil.java:33)
```

Since we now say autocorrect support is not available in detekt itself, but can be implemented in the rules or rule providers themselves, we could switch to using `KotlinCoreEnvironment.getSourceFiles()` and say that rule implementers have to create writeable versions of files themselves if they want to support autocorrect.

I personally think that's a reasonable tradeoff.

We can easily get rid of `FilePath` then :D

--- a/detekt-api/build.gradle.kts
+++ b/detekt-api/build.gradle.kts
@@ -7,7 +7,6 @@ plugins {
 
 dependencies {
     api(libs.kotlin.compilerEmbeddable)
-    api(projects.detektPsiUtils)
     implementation(projects.detektUtils)
 
     testImplementation(projects.detektTest)

--- a/detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Location.kt
+++ b/detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Location.kt
@@ -1,7 +1,6 @@
 package io.gitlab.arturbosch.detekt.api
 
 import dev.drewhamilton.poko.Poko
-import io.github.detekt.psi.absolutePath
 import org.jetbrains.kotlin.com.intellij.openapi.util.TextRange
 import org.jetbrains.kotlin.com.intellij.psi.PsiElement
 import org.jetbrains.kotlin.diagnostics.DiagnosticUtils.getLineAndColumnInPsiFile
@@ -10,6 +9,7 @@ import org.jetbrains.kotlin.psi.KtFile
 import org.jetbrains.kotlin.psi.psiUtil.endOffset
 import org.jetbrains.kotlin.psi.psiUtil.startOffset
 import java.nio.file.Path
+import kotlin.io.path.Path
 
 /**
  * Specifies a position within a source code fragment.
@@ -37,7 +37,12 @@ class Location(
             val end = endLineAndColumn(element, offset)
             val endSourceLocation = SourceLocation(end.line, end.column)
             val textLocation = TextLocation(element.startOffset + offset, element.endOffset + offset)
-            return Location(sourceLocation, endSourceLocation, textLocation, element.containingFile.absolutePath())
+            return Location(
+                sourceLocation,
+                endSourceLocation,
+                textLocation,
+                Path((element.containingFile as KtFile).virtualFilePath)
+            )
         }
 
         /**

--- a/detekt-formatting/build.gradle.kts
+++ b/detekt-formatting/build.gradle.kts
@@ -6,6 +6,7 @@ val extraDepsToPackage: Configuration by configurations.creating
 
 dependencies {
     compileOnly(projects.detektApi)
+    compileOnly(projects.detektPsiUtils)
     implementation(libs.ktlintRulesetStandard) {
         exclude(group = "org.jetbrains.kotlin")
     }

--- a/detekt-generator/build.gradle.kts
+++ b/detekt-generator/build.gradle.kts
@@ -16,6 +16,7 @@ val generatedUsageOutput by configurations.resolvable("generatedUsageOutput") {
 dependencies {
     implementation(projects.detektParser)
     implementation(projects.detektApi)
+    implementation(projects.detektPsiUtils)
     generatedUsage(projects.detektCli) {
         targetConfiguration = "generatedCliUsage"
     }

--- a/detekt-metrics/build.gradle.kts
+++ b/detekt-metrics/build.gradle.kts
@@ -4,7 +4,9 @@ plugins {
 
 dependencies {
     api(projects.detektApi)
+    compileOnly(projects.detektPsiUtils)
     testImplementation(projects.detektTestUtils)
     testImplementation(testFixtures(projects.detektApi))
     testImplementation(libs.assertj)
+    testRuntimeOnly(projects.detektPsiUtils)
 }

--- a/detekt-rules-complexity/build.gradle.kts
+++ b/detekt-rules-complexity/build.gradle.kts
@@ -5,6 +5,7 @@ plugins {
 dependencies {
     compileOnly(projects.detektApi)
     compileOnly(projects.detektMetrics)
+    compileOnly(projects.detektPsiUtils)
     testImplementation(projects.detektMetrics)
     testImplementation(projects.detektTest)
     testImplementation(libs.assertj)

--- a/detekt-rules-coroutines/build.gradle.kts
+++ b/detekt-rules-coroutines/build.gradle.kts
@@ -4,6 +4,7 @@ plugins {
 
 dependencies {
     compileOnly(projects.detektApi)
+    compileOnly(projects.detektPsiUtils)
     testImplementation(projects.detektTest)
     testImplementation(libs.assertj)
     testRuntimeOnly(libs.kotlinx.coroutines)

--- a/detekt-rules-documentation/build.gradle.kts
+++ b/detekt-rules-documentation/build.gradle.kts
@@ -4,6 +4,7 @@ plugins {
 
 dependencies {
     compileOnly(projects.detektApi)
+    compileOnly(projects.detektPsiUtils)
     testImplementation(projects.detektTest)
     testImplementation(libs.assertj)
 }

--- a/detekt-rules-empty/build.gradle.kts
+++ b/detekt-rules-empty/build.gradle.kts
@@ -4,6 +4,7 @@ plugins {
 
 dependencies {
     compileOnly(projects.detektApi)
+    compileOnly(projects.detektPsiUtils)
     testImplementation(projects.detektTest)
     testImplementation(libs.assertj)
 }

--- a/detekt-rules-errorprone/build.gradle.kts
+++ b/detekt-rules-errorprone/build.gradle.kts
@@ -4,6 +4,7 @@ plugins {
 
 dependencies {
     compileOnly(projects.detektApi)
+    compileOnly(projects.detektPsiUtils)
     implementation(projects.detektTooling)
     testImplementation(projects.detektTest)
     testImplementation(libs.assertj)

--- a/detekt-rules-exceptions/build.gradle.kts
+++ b/detekt-rules-exceptions/build.gradle.kts
@@ -4,6 +4,7 @@ plugins {
 
 dependencies {
     compileOnly(projects.detektApi)
+    compileOnly(projects.detektPsiUtils)
     testImplementation(projects.detektTest)
     testImplementation(libs.assertj)
 }

--- a/detekt-rules-libraries/build.gradle.kts
+++ b/detekt-rules-libraries/build.gradle.kts
@@ -5,6 +5,7 @@ plugins {
 dependencies {
     compileOnly(kotlin("stdlib-jdk8"))
     compileOnly(projects.detektApi)
+    compileOnly(projects.detektPsiUtils)
     testImplementation(projects.detektTest)
     testImplementation(libs.assertj)
 }

--- a/detekt-rules-naming/build.gradle.kts
+++ b/detekt-rules-naming/build.gradle.kts
@@ -4,6 +4,7 @@ plugins {
 
 dependencies {
     compileOnly(projects.detektApi)
+    compileOnly(projects.detektPsiUtils)
     testImplementation(projects.detektTest)
     testImplementation(libs.assertj)
 }

--- a/detekt-rules-ruleauthors/build.gradle.kts
+++ b/detekt-rules-ruleauthors/build.gradle.kts
@@ -5,6 +5,7 @@ plugins {
 dependencies {
     compileOnly(kotlin("stdlib-jdk8"))
     compileOnly(projects.detektApi)
+    compileOnly(projects.detektPsiUtils)
     testImplementation(projects.detektTest)
     testImplementation(libs.assertj)
 }

--- a/detekt-rules-style/build.gradle.kts
+++ b/detekt-rules-style/build.gradle.kts
@@ -5,6 +5,7 @@ plugins {
 dependencies {
     compileOnly(projects.detektApi)
     compileOnly(projects.detektMetrics)
+    compileOnly(projects.detektPsiUtils)
     testImplementation(projects.detektMetrics)
     testImplementation(projects.detektTest)
     testImplementation(libs.assertj)

